Add Preferences search (fix #5353)

ckaiser opened the pull request:
Fixes #5353, adds a search to the preferences menu like so:

(**Note**: Outdated video)

https://github.com/user-attachments/assets/8c292720-b55c-4edf-85d8-6e5a1064cc3c

ckaiser commented:
Fixed some issues with the disabling/re-enabling and added a disabled state to the `ColorButton` basically making it greyscale, since it had no disabled feedback.

@dacap Made a change to Slider so it's also setting the `text()` to the value, I think it makes sense semantically but I could also just work around that in my code and not change a core class.

Gasparoken commented:
Almost perfect, but...:
```cpp
          if (label->isEnabled()) {
            label->buddy()->setEnabled(true);

            // Special case to handle buttonsets
            if (label->buddy()->type() == kGridWidget) {
              for (auto* child : label->buddy()->children())
                child->setEnabled(true);
            }
          }
          else if (label->buddy()->isEnabled() && !label->buddy()->text().empty()) {
            label->setEnabled(true);
          }
```
I found that the ComboBox labels are disabled regardless, even though the string "pn" belongs to one of the available options ("png") in the default file association ComboBox under "Files" section.

The problem lies in the following "else if" statement:
```cpp
  else if (label->buddy()->isEnabled() && !label->buddy()->text().empty())
```
In the mentioned ComboBox `label->buddy()->isEnabled()` is "true", however, `!label->buddy()->text().empty()` is "false" since `!label->buddy()` is a ComboBox type with no text. Finally, in the UI, the ComboBox label will incorrectly change to a disabled state.

ckaiser commented:
@Gasparoken Fixed! It was old code that I forgot to take out when I changed the approach to combo boxes 🤦

Gasparoken commented:
I think it's fine now!
Just one observation for the future: When I search for "co", the "Colors:" label in the "Background" section has disabled buttons, but they should be enabled. This happens because the "Colors:" label doesn't have a specific "buddy" (it's in nullptr). I see that this could be resolved with one of two alternatives:
1. My preferred solution would be to allow "Labels" to have a "Grid" as their "buddy", so `label->buddy()` could return an array with all the displayable widgets.
2. Address this specific case in some way, since it only happens in this instance.

In any case, I think it could be mixed in as is, and a separate issue could be opened for this detail.

ckaiser commented:
Yeah, ideally I think we should be searching and highlighting "rows" of widgets but that required a way to tag them or completely reorganize the entire XML somehow to get that to work, so I went for the approach that had the least friction with the current code and layout. I'll see about a way to either set the buddy for this or handle the case specifically.

ckaiser commented:
### Latest polish pass

 * Adjusted the color buttons so the text would also be slightly transparent to more accurately portray the fact it's disabled
 * Added a `setEnabled` call to correspond to most of the other `setVisible` calls in the code, this way we're not accidentally searching inside things like the font warning, OS things that don't exist for the current platform, etc.
 * Added a `resetSearch` function and started calling on some disruptive events like things that spawn extra controls (custom cursor color, etc).
 * Added exceptions for the new document/active document combo boxes.
 * Added some hboxes with IDs for things that are closely connected, this way if you search for the label, you'll be able to see all the relevant things highlighted instead of just the first one, for example:
<img width="871" height="220" alt="image" src="https://github.com/user-attachments/assets/f90ada54-83e2-472d-947f-eaae8b1f7493" />


### Related to this
@dacap I found some weird behavior regarding the checkered background size option, presumably complicated by the logic of it having to switch between active or new documents. Right now you can't change this option and then switch to another section without losing what you did, which feels really inconsistent with other options, this is happening in the current stable release, figured I'd mention it, seems like the logic handling this could use a refactor.

dacap commented:
Found this bug giving a try (cannot change font after a search and canceling the search):

![search-pref-bug](https://github.com/user-attachments/assets/c39fefd5-4b9e-4be7-b272-60388124de94)

I've also noticed that we have to cancel the search to enable some item that we might click. Probably something to see in the future how we could fix it (e.g. just clicking an element that is "temporarily disabled" cancels the search / or re-enable all items in the section).

dacap commented:
> @dacap I found some weird behavior regarding the checkered background size option, presumably complicated by the logic of it having to switch between active or new documents. Right now you can't change this option and then switch to another section without losing what you did, which feels really inconsistent with other options, this is happening in the current stable release, figured I'd mention it, seems like the logic handling this could use a refactor.

We can leave a fix for the future, probably with a new "Set as default" option, but it's other issue.

ckaiser commented:
Fixed the weird issues, it was a byproduct of me relying on `isEnabled()` instead of using the raw flags, now things work more smoothly.

With regards to what happens when you click on things, my first solution was to just reset the search manually in all of the different `onSomething[..]()` functions, for things that might have a potential side effect, but that presents too many limitations and weird quirks, so my new solution is a filter where any focus change that isn't to the section list or the search box will just clear the search and reset things to normal. This works with mouse and keyboard selection as well.

Presumably by focusing on something specific the user has finished searching and this works pretty well because the focus message runs before anything else

Other changes:
 * Simplified code a fair bit
 * Now using the fuzzy search from the runner (although a bit more strict)
 * We can search widget tooltips (@dacap  This PR adds a way for generated widgets to access tooltip managers, figured this was the cleanest way to implement it, lmk what you think).
 * Labels are better at finding their buddies, and at asserting when they can't (this was the cause of some issues)

dacap commented on data/widgets/keyboard_shortcuts.xml:
```diff
@@ -7,7 +7,7 @@
                 by="pixel" position="80" id="section_splitter">
         <vbox expansive="true">
-          <search id="search" magnet="true" />
+          <search id="search" magnet="true" clearonesc="true" debounce="250" />(
```
An extra `(`, it looks like a typo
```suggestion
          <search id="search" magnet="true" clearonesc="true" debounce="250" />
```
